[CI] SharedClusterSnapshotRestoreIT.testSnapshotSucceedsAfterSnapshotFailure fails

This test reminds me of older SharedClusterSnapshotRestoreIT failure on CI, but what is unique about
this one is that it just recently popped up. Although it has only been encountered twice in CI in the past 90 day, those two times were this week.

link to CI:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-unix-compatibility/os=debian/1047/console

reproduce step:

```
./gradlew :server:integTest -Dtests.seed=A207A05308BD2693 -Dtests.class=org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT -Dtests.method="testSnapshotSucceedsAfterSnapshotFailure" -Dtests.security.manager=true -Dtests.locale=uk-UA -Dtests.timezone=Europe/Belfast
```

stacktrace:

```
FAILURE 0.22s J1 | SharedClusterSnapshotRestoreIT.testSnapshotSucceedsAfterSnapshotFailure <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: expected:<0> but was:<1>
   > 	at __randomizedtesting.SeedInfo.seed([A207A05308BD2693:DD692EFC33F4422E]:0)
   > 	at org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT.testSnapshotSucceedsAfterSnapshotFailure(SharedClusterSnapshotRestoreIT.java:3150)
   > 	at java.lang.Thread.run(Thread.java:748)
  1> [2018-05-10T01:13:14,639][INFO ][o.e.s.SharedClusterSnapshotRestoreIT] [testSnapshotStatusOnFailedIndex]: before test
  1> [2018-05-10T01:13:14,639][INFO ][o.e.s.SharedClusterSnapshotRestoreIT] [SharedClusterSnapshotRestoreIT#testSnapshotStatusOnFailedIndex]: setting up test
  1> [2018-05-10T01:13:14,646][INFO ][o.e.c.m.MetaDataIndexTemplateService] [node_s0] adding template [random_index_template] for index patterns [*]
```

@tlrx I've quickly looked at these failures and I think it's caused by #30332. As we now first write the new index shard snapshots file, a previous (older) failed attempt might block us from doing so:

```
Caused by: java.nio.file.FileAlreadyExistsException: blob [pending-index-0] already exists, cannot overwrite
05:26:19   1> 	at org.elasticsearch.common.blobstore.fs.FsBlobContainer.writeBlob(FsBlobContainer.java:127) ~[main/:?]
05:26:19   1> 	at org.elasticsearch.snapshots.mockstore.BlobContainerWrapper.writeBlob(BlobContainerWrapper.java:53) ~[test/:?]
05:26:19   1> 	at org.elasticsearch.snapshots.mockstore.MockRepository$MockBlobStore$MockBlobContainer.writeBlob(MockRepository.java:361) ~[test/:?]
05:26:19   1> 	at org.elasticsearch.repositories.blobstore.ChecksumBlobStoreFormat.writeBlob(ChecksumBlobStoreFormat.java:191) ~[main/:?]
05:26:19   1> 	at org.elasticsearch.repositories.blobstore.ChecksumBlobStoreFormat.writeAtomic(ChecksumBlobStoreFormat.java:136) ~[main/:?]
05:26:19   1> 	at org.elasticsearch.repositories.blobstore.BlobStoreRepository$Context.finalize(BlobStoreRepository.java:955) ~[main/:?]
05:26:19   1> 	... 10 more
```

This looks serious to me, and I think we should fix this asap before it gets released. I see two options:
1) similar to before, start deleting tmp files (`indexShardSnapshotsFormat.isTempBlobName(blobName)`) before writing the new index file.
2) Adapt `ChecksumBlobStoreFormat.writeAtomic` to use a temp blob name that has a uuid in it.

--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
@@ -950,6 +950,20 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
 
             final BlobStoreIndexShardSnapshots updatedSnapshots = new BlobStoreIndexShardSnapshots(snapshots);
             try {
+                // Delete temporary index files first, as we might otherwise fail in the next step creating the new index file if an earlier
+                // attempt to write an index file with this generation failed mid-way after creating the temporary file.
+                for (final String blobName : blobs.keySet()) {
+                    if (indexShardSnapshotsFormat.isTempBlobName(blobName)) {
+                        try {
+                            blobContainer.deleteBlobIgnoringIfNotExists(blobName);
+                        } catch (IOException e) {
+                            logger.warn(() -> new ParameterizedMessage("[{}][{}] failed to delete index blob [{}] during finalization",
+                                snapshotId, shardId, blobName), e);
+                            throw e;
+                        }
+                    }
+                }
+
                 // If we deleted all snapshots, we don't need to create a new index file
                 if (snapshots.size() > 0) {
                     indexShardSnapshotsFormat.writeAtomic(updatedSnapshots, blobContainer, indexGeneration);
@@ -957,7 +971,7 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
 
                 // Delete old index files
                 for (final String blobName : blobs.keySet()) {
-                    if (indexShardSnapshotsFormat.isTempBlobName(blobName) || blobName.startsWith(SNAPSHOT_INDEX_PREFIX)) {
+                    if (blobName.startsWith(SNAPSHOT_INDEX_PREFIX)) {
                         try {
                             blobContainer.deleteBlobIgnoringIfNotExists(blobName);
                         } catch (IOException e) {

--- a/server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java
+++ b/server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java
@@ -3094,7 +3094,6 @@ public class SharedClusterSnapshotRestoreIT extends AbstractSnapshotIntegTestCas
         assertEquals("IndexShardSnapshotFailedException[Aborted]", snapshotInfo.shardFailures().get(0).reason());
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/30507")
     public void testSnapshotSucceedsAfterSnapshotFailure() throws Exception {
         logger.info("--> creating repository");
         final Path repoPath = randomRepoPath();

